### PR TITLE
Incremental development for UDP support

### PR DESCRIPTION
**Pull request by ammarfaizi2**
Hi,

This is incremental development for UDP support, I will batch the next
commits to dev_001 branch again.

```
----------------------------------------------------------------
The following changes since commit 30479d4c51b2e682563e955ef90af4fd933038fb:

  git: resolve merge conflict (2021-08-29 23:10:46 +0700)

are available in the Git repository at:

  https://github.com/ammarfaizi2/teavpn2-fork udp_support_for_dev_001-2021-19-02

for you to fetch changes up to d6b3badb8c58921ddb4b0566bc5cbad1a25044ce:

  client: address clang warnings (2021-09-02 20:06:00 +0700)

----------------------------------------------------------------
Ammar Faizi (14):
      Resolve merge conflict
      mutex.h: implement locking wrapper for extra assertions
      server: refactor UDP session management
      packet.h: new packet type TSRV_PKT_HANDSHAKE_REJECT
      {server,client}: implement handshake validation
      Merge branch 'udp_support'
      udp_session.c: delete unused comments
      udp.h: add prep functions to simplify packet construction
      udp_session.c: implement put_udp_session
      udp_session.c: fix potential read from uninitialized variable
      udp_epoll.c: add reap_zombie_sessions
      client: implement prep functions to simplify packet assignment
      server: interrupt the subthreads with SIGTERM upon exit
      client: address clang warnings

 Makefile                               |   5 +--
 src/teavpn2/client/linux/udp.c         |  84 ++++++++++++++++++++++++++++++---------------
 src/teavpn2/client/linux/udp.h         |  43 +++++++++++++++++++++++
 src/teavpn2/mutex.h                    | 108 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/teavpn2/packet.h                   |  33 ++++++++++++------
 src/teavpn2/server/linux/udp.c         |  81 ++++++++++++++++++++++++++++++++++----------
 src/teavpn2/server/linux/udp.h         | 118 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------
 src/teavpn2/server/linux/udp_epoll.c   | 271 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++--------------------------
 src/teavpn2/server/linux/udp_session.c | 188 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------
 src/teavpn2/stack.h                    |   9 +++++
 10 files changed, 765 insertions(+), 175 deletions(-)
 create mode 100644 src/teavpn2/mutex.h
```

### CODE REVIEW

**Comment by ammarfaizi2** [comment]
@louvian or anybody, want to review my code?

**Comment by louvian** [comment]
> @louvian or anybody, want to review my code?

Give me some time for it!

File: Makefile [hunk 1]
```diff
@@ -44,7 +44,7 @@ LDFLAGS		:= -ggdb3 -rdynamic
 
 # Flag to link any library to $(TARGET_BIN)
 # (end arguments)
-LIB_LDFLAGS	:= -lpthread
+LIB_LDFLAGS	:= -lasan -lpthread
```
Comment (louvian) [inline Makefile:47]: This `-lasan` is not conditional, so you compile with ASAN even in release mode?

File: src/teavpn2/server/linux/udp_epoll.c [hunk 2]
```diff
@@ -399,15 +517,16 @@ static int do_epoll_wait(struct epl_thread *thread)
...
 	events = thread->events;
 	for (i = 0; i < ret; i++) {
 		tmp = handle_event(thread, &events[i]);
 		if (unlikely(tmp))
 			return tmp;
 	}
 
+	if (thread->idx > 0)
+		reap_zombie_sessions(thread);
+
```
[the remark this answers is not in the document]
Comment (ammarfaizi2) [reply src/teavpn2/server/linux/udp_epoll.c:529]: Yes, you are right, I missed it.

BTW, I want to create one additional thread for zombie reaper, this thread will be low priority, what do you think?

[the remark this answers is not in the document]
Comment (ammarfaizi2) [reply src/teavpn2/server/linux/udp_epoll.c:529]: I don't want to bother the busy threads which are reading/receiving data.

File: Makefile (hunk 1 above)
Comment (ammarfaizi2) [reply Makefile:47]: Oh, I should make it debug only flag, will address it later.

File: src/teavpn2/server/linux/udp_epoll.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (louvian) [reply src/teavpn2/server/linux/udp_epoll.c:529]: > BTW, I want to create one additional thread for zombie reaper, this thread will be low priority, what do you think?

I agree. Don't see the badness so far.

[the remark this answers is not in the document]
Comment (ammarfaizi2) [reply src/teavpn2/server/linux/udp_epoll.c:529]: Alright, that's all for today. Thanks for responding my request.

**Comment by ammarfaizi2** [comment]
I will spin more patches to get it ready (hopefully ready for testing tonight).

**Comment by ammarfaizi2** [comment]
@ammarfaizi2 I am delaying this, let's see if I have the time for it tomorrow.

**Comment by ammarfaizi2** [comment]
Working on this...

**Comment by ammarfaizi2** [comment]
Wasn't too bad for the first try, 4K ping flood with teavpn2 debug mode:
```
$ sudo ping -f 10.1.1.1 -s 4096
PING 10.1.1.1 (10.1.1.1) 4096(4124) bytes of data.
......^C                 
--- 10.1.1.1 ping statistics ---
946 packets transmitted, 940 received, 0.634249% packet loss, time 15895ms
rtt min/avg/max/mdev = 47.219/118.896/383.700/59.467 ms, pipe 22, ipg/ewma 16.820/81.145 ms
```

**Comment by ammarfaizi2** [comment]
I got bad length log from tcpdump here, I guess I should fix the MTU.

```
21:47:42.689434 IP 10.7.7.2.47962 > 68.183.184.174.44444: UDP, bad length 1480 > 1472
21:47:42.689626 IP 10.7.7.2.47962 > 68.183.184.174.44444: UDP, bad length 1480 > 1472
```

**Comment by ammarfaizi2** [comment]
It says `bad length 1480 > 1472`, so I should cap the transfer bytes to 1472 per `sendto()` call.

**Comment by ammarfaizi2** [comment]
Let me have one extra day for fixing this...

**Comment by ammarfaizi2** [comment]
Good morning.

This is a simple test after a bit fixing, still under debug mode with `-O0` compile flag and high level of log verbosity (with no CPU affinity utilization). I am confident this can be further optimized.

Screenshot_2021-09-06_05-20-35

Will take a look into it further after work.

**Comment by ammarfaizi2** [comment]
Fixed, I will prepare the setup instructions for testers tonight 🎉

The part that I doubt is:
Can my code handle multiple clients?
^ Let's see the test result later.

Extra: I will require you to run at least Linux 5.10+ for this.

**Comment by ammarfaizi2** [comment]
New diffstat here:

```
----------------------------------------------------------------
The following changes since commit 30479d4c51b2e682563e955ef90af4fd933038fb:

  git: resolve merge conflict (2021-08-29 23:10:46 +0700)

are available in the Git repository at:

  https://github.com/ammarfaizi2/teavpn2-fork udp_support_for_dev_001-2021-19-02

for you to fetch changes up to fefdf802173f3685c5439ba731731ff952053f62:

  Merge branch 'udp_support' into udp_support_for_dev_001-2021-19-02 (2021-09-06 18:42:28 +0700)

----------------------------------------------------------------
Ammar Faizi (30):
      Resolve merge conflict
      mutex.h: implement locking wrapper for extra assertions
      server: refactor UDP session management
      packet.h: new packet type TSRV_PKT_HANDSHAKE_REJECT
      {server,client}: implement handshake validation
      Merge branch 'udp_support'
      udp_session.c: delete unused comments
      udp.h: add prep functions to simplify packet construction
      udp_session.c: implement put_udp_session
      udp_session.c: fix potential read from uninitialized variable
      udp_epoll.c: add reap_zombie_sessions
      client: implement prep functions to simplify packet assignment
      server: interrupt the subthreads with SIGTERM upon exit
      client: address clang warnings
      auth.c: add teavpn2_auth function
      {client,server}: implement auth and auth response
      all: clean up
      README.md: update code of conduct
      {client,server}: fix packet routing
      {client,server}: clean up
      server/linux/udp_epoll.c: change zombie reaper verbosity level to 6
      config: fix up config and adjust default MTU
      client: fix override_default option behavior
      {client,server}: increase default buffer size to reduce emergency soft loop
      server: close all clients before exit
      client: fix auth fail error message
      server: don't treat repeated close packet as new connection
      udp_epoll.c: trivial semicolon fix (syntax error)
      server: skip init for several packet types
      Merge branch 'udp_support' into udp_support_for_dev_001-2021-19-02

 README.md                              | 146 ++++++++++++++++++++++-----------
 config/client.ini                      |  11 ++-
 config/server.ini                      |   6 +-
 data/server/users/ammarfaizi2.ini      |   6 +-
 src/ext/Makefile                       |   5 +-
 src/teavpn2/Makefile                   |   1 +
 src/teavpn2/auth.c                     | 155 +++++++++++++++++++++++++++++++++++
 src/teavpn2/client/linux/udp.c         | 250 ++++++++++++++++++++++++++++++++++++++++++++++++++-------
 src/teavpn2/client/linux/udp.h         |  49 ++++++++++++
 src/teavpn2/client/linux/udp_epoll.c   | 348 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---
 src/teavpn2/common.h                   |  26 ++++++
 src/teavpn2/mutex.h                    | 108 +++++++++++++++++++++++++
 src/teavpn2/packet.h                   |  41 +++++++---
 src/teavpn2/print.c                    |  15 ++++
 src/teavpn2/print.h                    |   8 ++
 src/teavpn2/server/entry.c             |  39 ++++-----
 src/teavpn2/server/linux/udp.c         | 161 ++++++++++++++++++++++++++++++++-----
 src/teavpn2/server/linux/udp.h         | 184 ++++++++++++++++++++++++++++++++++++++----
 src/teavpn2/server/linux/udp_epoll.c   | 640 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---------------
 src/teavpn2/server/linux/udp_session.c | 189 ++++++++++++++++++++++++++++++-------------
 src/teavpn2/stack.h                    |   7 ++
 21 files changed, 2107 insertions(+), 288 deletions(-)
 create mode 100644 src/teavpn2/auth.c
 create mode 100644 src/teavpn2/mutex.h
```

**Comment by ammarfaizi2** [comment]
@louvian are you free tonight?

File: Makefile (hunk 1 above)
Comment (ammarfaizi2) [reply Makefile]: Addressed, thanks for the feedback!

**Comment by louvian** [comment]
On Mon, Sep 6, 2021 at 7:19 PM Ammar Faizi ***@***.***> wrote:
>
> @louvian are you free tonight?
>

Sorry Ammar, I can't test it tonight, I only bring my phone. I'm at
the hospital taking care of my mom because she is sick. I will mail
you when I am home.

**Comment by ammarfaizi2** [comment]
> Sorry Ammar, I can't test it tonight, I only bring my phone. I'm at
> the hospital taking care of my mom because she is sick. I will mail
> you when I am home.

Uh oh, I am sorry to hear that. Yes, please take care of her with all your best effort, she is absolutely more important than TeaVPN2. All the best wishes to you and your mother. I hope she will get well soon :)

**Comment by ammarfaizi2** [comment]
Fixes: https://github.com/TeaInside/teavpn2/issues/3

**Pull request merged**